File: YelpReviews/app/src/main/java/com/somethingsimple/yelpreviews/contracts/Callback.java
package com.somethingsimple.yelpreviews.contracts;

/**
 * Callback interface to provide a contract fr service layer to provide a callback with data.
 */
public interface Callback<T> {

    /**
     * success callback
     * @param response success response model.
     */
    void onSuccess(T response);

    /**
     * failure callback
     */
    void onFailure();

    /**
     * Returns class type of model.
     * @return model class type.
     */
    Class<T> getClassType();
}

File: Kobo/KoboChallenge/app/src/main/java/com/kobo/demo/challenge/network/Callback.kt
package com.kobo.demo.challenge.network

/**
 * Callback interface to provide a contract fr service layer to provide a callback with data.
 */
interface Callback<T> {

    /**
     * Returns class type of model.
     * @return model class type.
     */
    val classType: Class<T>

    /**
     * success callback
     * @param response success response model.
     */
    fun onSuccess(response: T)

    /**
     * failure callback
     */
    fun onFailure()
}
